File: protocol/src/test/lua/LuaProtocol/ProtocolManager.lua
protocols = {}

ProtocolManager = {}

-- table扩展方法，map的大小
function table.mapSize(map)
    local size = 0
    for _,_ in pairs(map) do
        size = size + 1
    end
    return size
end

function ProtocolManager.getProtocol(protocolId)
    local protocol = protocols[protocolId]
    if protocol == nil then
        error("[protocolId:" + protocolId + "]协议不存在")
    end
    return protocol
end

function ProtocolManager.write(buffer, packet)
    local protocolId = packet:protocolId()
    -- 写入协议号
    buffer:writeShort(protocolId)
    -- 写入包体
    ProtocolManager.getProtocol(protocolId):write(buffer, packet)
end

function ProtocolManager.read(buffer)
    local protocolId = buffer:readShort()
    return ProtocolManager.getProtocol(protocolId):read(buffer)
end

function initProtocol()
    local ComplexObject = require("LuaProtocol.Packet.ComplexObject")
    local NormalObject = require("LuaProtocol.Packet.NormalObject")
    local ObjectA = require("LuaProtocol.Packet.ObjectA")
    local ObjectB = require("LuaProtocol.Packet.ObjectB")
    local SimpleObject = require("LuaProtocol.Packet.SimpleObject")
    protocols[100] = ComplexObject
    protocols[101] = NormalObject
    protocols[102] = ObjectA
    protocols[103] = ObjectB
    protocols[104] = SimpleObject
end

ProtocolManager.initProtocol = initProtocol
return ProtocolManager

File: protocol/src/test/ecmascript/zfooes/ProtocolManager.mjs
import EmptyObject from './packet/EmptyObject.mjs';
import { EmptyObjectRegistration } from './packet/EmptyObject.mjs';
import VeryBigObject from './packet/VeryBigObject.mjs';
import { VeryBigObjectRegistration } from './packet/VeryBigObject.mjs';
import ComplexObject from './packet/ComplexObject.mjs';
import { ComplexObjectRegistration } from './packet/ComplexObject.mjs';
import NormalObject from './packet/NormalObject.mjs';
import { NormalObjectRegistration } from './packet/NormalObject.mjs';
import ObjectA from './packet/ObjectA.mjs';
import { ObjectARegistration } from './packet/ObjectA.mjs';
import ObjectB from './packet/ObjectB.mjs';
import { ObjectBRegistration } from './packet/ObjectB.mjs';
import SimpleObject from './packet/SimpleObject.mjs';
import { SimpleObjectRegistration } from './packet/SimpleObject.mjs';

const protocols = new Map();
const protocolIdMap = new Map();

// initProtocol
protocols.set(0, new EmptyObjectRegistration());
protocolIdMap.set(EmptyObject, 0);
protocols.set(1, new VeryBigObjectRegistration());
protocolIdMap.set(VeryBigObject, 1);
protocols.set(100, new ComplexObjectRegistration());
protocolIdMap.set(ComplexObject, 100);
protocols.set(101, new NormalObjectRegistration());
protocolIdMap.set(NormalObject, 101);
protocols.set(102, new ObjectARegistration());
protocolIdMap.set(ObjectA, 102);
protocols.set(103, new ObjectBRegistration());
protocolIdMap.set(ObjectB, 103);
protocols.set(104, new SimpleObjectRegistration());
protocolIdMap.set(SimpleObject, 104);

class ProtocolManager {
    static getProtocolId(clazz) {
        const protocolId = protocolIdMap.get(clazz);
        if (protocolId === null || protocolId === undefined) {
            throw '[protocol:' + clazz + '] not exist';
        }
        return protocolId;
    }
    static getProtocol(protocolId) {
        const protocol = protocols.get(protocolId);
        if (protocol === null) {
            throw new Error('[protocolId:' + protocolId + '] not exist');
        }
        return protocol;
    }

    static write(buffer, packet) {
        const protocolId = ProtocolManager.getProtocolId(packet.constructor);
        buffer.writeShort(protocolId);
        const protocol = ProtocolManager.getProtocol(protocolId);
        protocol.write(buffer, packet);
    }

    static read(buffer) {
        const protocolId = buffer.readShort();
        const protocol = ProtocolManager.getProtocol(protocolId);
        const packet = protocol.read(buffer);
        return packet;
    }
}


export default ProtocolManager;